{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "3dcee117-1ec6-46b0-8131-3012d913baf0",
   "metadata": {},
   "source": [
    "## 1. Preparation\n",
    "\n",
    "### 1.1 Library the packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5c9c3b00-5c7a-4ce8-9b2f-7516281090bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(tidyverse)\n",
    "library(vegan)\n",
    "library(ape)\n",
    "library(microbiome)\n",
    "library(mediation)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "92048656-39a7-4be0-bd51-66992e163455",
   "metadata": {},
   "source": [
    "### 1.2 Load the datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5f9ca7ce-a9a1-4307-b8de-6b1a173aa6c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "mbio <- read.table(\"/data/precomputed/lld.metaphlan2.species.tsv\") # Species abundance table\n",
    "phen <- read.table(\"/data/precomputed/lld.phen.tsv\")               # Phenotype table"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "567d6a60-188a-400c-b907-98115b30fa45",
   "metadata": {},
   "source": [
    "## 2 Preprocess\n",
    "\n",
    "The original metaphlan2 output looks 'ugly', we can make some modification to make it nicer."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "30b68621-ef2b-4e6b-bd7a-8ffcd2331074",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Shorten the long taxonomy names, only keep the species names\n",
    "rownames(mbio) <- str_replace_all(rownames(mbio), \".*s__\", \"s__\")\n",
    "\n",
    "# transpose the abundance table: rows are samples and columns are species\n",
    "mbio <- t(mbio)\n",
    "\n",
    "# Rescale the table, make sure the sum of each row is 1\n",
    "mbio <- apply(mbio, 1, myfun<-function(x){x/sum(x)}) %>% t() %>% as.data.frame()\n",
    "# rowSums(mbio) # Check the result, all 1 is correct."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c672eb00-6870-44ee-9258-085dce597f95",
   "metadata": {},
   "source": [
    "## 3 Overall ecological parameters\n",
    "\n",
    "For the analysis of overall ecological parameters (alpha diversity, beta diversity...), we should use the **unfiltered** abundance table.\n",
    "\n",
    "### 3.1 Alpha diversity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1a610e88-66c0-4f81-8800-6651928157d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate Shannon diversity using vegan's function, and count the species number (richness).\n",
    "mbio.alpha <- data.frame(Shannon  = vegan::diversity(mbio, index=\"shannon\"), \n",
    "                         Richness = rowSums(mbio!=0))\n",
    "\n",
    "if(!dir.exists(\"alphaDiversity\")){dir.create(\"alphaDiversity\")}\n",
    "write.table(mbio.alpha, \"alphaDiversity/alphaDiversity.tsv\", sep = \"\\t\", row.names = T, col.names = NA, quote = F)\n",
    "\n",
    "# Check if there is a difference in gut microbiome Shonnon index between hyperglycemia patients and health controls\n",
    "wilcox.test(mbio.alpha$Shannon ~ phen$Hyperglycemia)\n",
    "\n",
    "# Visualize the between-group differences in Shannon index using a boxplot\n",
    "ggplot(mbio.alpha, aes(as.factor(phen$Hyperglycemia), Shannon, color = as.factor(phen$Hyperglycemia)))+\n",
    "  geom_boxplot() + \n",
    "  xlab(\"Groups\") +\n",
    "  ylab(\"Shannon index\") + \n",
    "  theme(legend.position = \"none\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cc57956b-8f65-467e-b173-ee3a4c288d15",
   "metadata": {},
   "source": [
    "### 3.2 Beta diversity\n",
    "\n",
    "Principal coordinates analysis (PCoA) can help us reduce the dimension of microbiome composition data, capture the main information of high-dimension dataset and project them to a few number of new dimensions. The PCoA plot can visually reflect the distance (dissimilarity) between samples."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b277442b-2d7d-40da-b3cd-feabfca72b47",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate dissimilarities in microbiome composition between samples.\n",
    "mbio.dist <- vegdist(mbio, index = \"bray\")\n",
    "\n",
    "# PCoA\n",
    "mbio.pcoa <- ape::pcoa(mbio.dist)\n",
    "\n",
    "# Extract the principal coordinates (PCs) (new coordinates generated by PCoA)\n",
    "mbio.pc   <- data.frame(mbio.pcoa$vectors)\n",
    "\n",
    "# PCoA plot: visualize the first 2 principal coordinates (PCs)\n",
    "ggplot(mbio.pc, aes(Axis.1, Axis.2, color = as.factor(phen$Hyperglycemia))) +\n",
    "  geom_point() + \n",
    "  xlab(\"PCo1\") +\n",
    "  ylab(\"PCo2\") +\n",
    "  theme(legend.position = \"none\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "685f780c-6e32-4ffa-9879-9c7c30098f05",
   "metadata": {},
   "source": [
    "### 3.3 PERMANOVA\n",
    "\n",
    "Permutational multivariate analysis of variance (PERMANOVA) can help us estimate how many proportion of the microbiome composition variance can be explained by each factor we care about."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7b890d75-437c-4cb5-aef0-71ed97ae8c6b",
   "metadata": {},
   "outputs": [],
   "source": [
    "mbio.adonis.res <- adonis2(mbio.dist ~ Age + Gender + FruitIntakeFrequency + Glucose, phen, permutations = 999)\n",
    "\n",
    "if(!dir.exists(\"betaDiversity\")){dir.create(\"betaDiversity\")}\n",
    "write.table(as.data.frame(mbio.adonis.res), \"betaDiversity/PERMANOVA.tsv\", sep = \"\\t\", row.names = T, col.names = NA, quote = F)\n",
    "knitr::kable(mbio.adonis.res)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9adb4780-f9e9-4fd2-a861-39b0ed289c2e",
   "metadata": {},
   "source": [
    "## 4 Preparation before association analysis\n",
    "\n",
    "### 4.1 Pre-calculation of prevalence and mean abundance\n",
    "\n",
    "Calculate the prevalence and mean abundance of each species, and get the list of species which are highly abundant and highly prevalent in our samples, then extract them **after CLR transformation**."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1f970a73-590b-4a5a-a765-3016929e4bd0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate the abundance and prevalence of each species\n",
    "spe.meanAbun <- colMeans(mbio)             # Mean abundance\n",
    "spe.nonZero  <- colSums(mbio>0)/nrow(mbio) # Prevalence\n",
    "\n",
    "# Get the list of species which mean abundance is greater than 0.01% and present in more than 5% samples\n",
    "# We will remove the species that are not in this list later\n",
    "spe.keep <- colnames(mbio)[spe.meanAbun > 0.0001 & spe.nonZero > 0.05]\n",
    "# length(spe.keep) # Check how many species are in the list."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c514e575-e44d-4548-9d82-109fc4b8f001",
   "metadata": {},
   "source": [
    "### 4.2 Centered log-ratio (clr) transformation and filtering\n",
    "\n",
    "Relative abundance profile is a compositional data, we remove the dependency between variables using CLR transformation, then do the filtering."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4e03a907-6633-4b6b-86c1-ca443232edc3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# CLR transformation, please note that the input table of microbiome::transform() function requires samples in columns and taxa in rows, this is why we transpose the table with the function t() and then transpose it back.\n",
    "mbio.clr <- microbiome::transform( t(mbio), transform = \"clr\") %>% t()\n",
    "\n",
    "# Filter out the species with a low abundance or low prevalence (or extract the highly abundant and highly prevalent species)\n",
    "mbio.clr.filtered <- mbio.clr[,spe.keep]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7feb67c8-4d3e-4e89-8fe9-ab5fcbd9c069",
   "metadata": {},
   "source": [
    "### 4.3 Assessment and removal of confounding effect caused by sequencing depth\n",
    "\n",
    "In metagenomic sequencing-based microbiome studies, sequencing depth always varies across different samples, and may bias the diversity estimation and species associations. Two strategies can be used to solve the sequencing depth issue, (1) **Rarefaction or downsize**, it means we randomly re-sample the reads from sequencing data for each sample to certain number of read count, in such a way all sample will have the same sequencing depth, but the drawback of this approach is that it wastes data and money. Another strategy is (2) **Including sequencing depth (read count) as a covariate** in downstream analysis, using statistical approach to remove the confounding effect introduced by sequencing depth. In our studies and this course, we recommend the second strategy.\n",
    "\n",
    "Here we can assess the impact of sequencing depth on microbiome diversity estimation and species abundance."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ed828fb1-9aba-4c55-9300-8e93ee488078",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check impact of sequencing depth on microbiome diversity\n",
    "cor.test(phen$CleanReadCount, mbio.alpha$Shannon)\n",
    "cor.test(phen$CleanReadCount, mbio.alpha$Richness)\n",
    "\n",
    "# Association between species abundance and read count\n",
    "spe.read.nocovar <- as.data.frame(matrix(NA, ncol = 3, nrow = ncol(mbio.clr.filtered)))\n",
    "colnames(spe.read.nocovar)<-c(\"Species\", \"R\", \"P\")\n",
    "\n",
    "for (i in 1:ncol(mbio.clr.filtered)) {\n",
    "    spe_i <- cor.test(mbio.clr.filtered[ ,i], phen$CleanReadCount, method = \"spearman\") # Spearman correlation analysis\n",
    "    \n",
    "    spe.read.nocovar$Species[i] <- colnames(mbio.clr.filtered)[i]             # Species name\n",
    "    spe.read.nocovar$R[i]       <- spe_i$estimate                             # R value (correlation strength) from Spearman correlation\n",
    "    spe.read.nocovar$P[i]       <- spe_i$p.value                              # P value from Spearman correlation\n",
    "}\n",
    "\n",
    "spe.read.nocovar$FDR <- p.adjust(spe.read.nocovar$P, method = \"fdr\") # Get adjusted P value\n",
    "spe.read.nocovar     <- dplyr::arrange(spe.read.nocovar, P)          # Sort result table by raw P value\n",
    "\n",
    "if(!dir.exists(\"beforeAssociation\")){dir.create(\"beforeAssociation\")}\n",
    "write.table(spe.read.nocovar, \"beforeAssociation/read.species.spearman.tsv\", sep = \"\\t\", row.names = F, col.names = T, quote = F)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8da4767d-0d0d-4fc3-8dfe-16571d81bfe6",
   "metadata": {},
   "source": [
    "## 5 Association analysis\n",
    "\n",
    "### 5.1 Correlation between species abundance and phenotypes (No covariate)\n",
    "\n",
    "We use **Spearman's correlation** (between two continuous variables) or **Wilcoxon rank-sum test** (between a continuous variable and a binary variable 0/1) to conduct association analysis if we don't consider any confounding effect."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0717976b-43eb-466e-a634-6f890c69d31b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Association between species abundance and binary variables (or between group comparison for binary variables)\n",
    "spe.hyperglycemia.nocovar <- as.data.frame(matrix(NA, ncol = 4, nrow = ncol(mbio.clr.filtered)))\n",
    "colnames(spe.hyperglycemia.nocovar)<-c(\"Species\", \"Mean_0\", \"Mean_1\", \"P\")\n",
    "\n",
    "for (i in 1:ncol(mbio.clr.filtered)) {\n",
    "    spe_i <- wilcox.test(mbio.clr.filtered[ ,i] ~ phen$Hyperglycemia)                         # Wilcoxon rank-sum test\n",
    "    \n",
    "    spe.hyperglycemia.nocovar$Species[i] <- colnames(mbio.clr.filtered)[i]                    # Species name\n",
    "    spe.hyperglycemia.nocovar$Mean_0[i]  <- mean(mbio.clr.filtered[phen$Hyperglycemia==0, i]) # Mean species abundance in group 0\n",
    "    spe.hyperglycemia.nocovar$Mean_1[i]  <- mean(mbio.clr.filtered[phen$Hyperglycemia==1, i]) # Mean species abundance in group 1\n",
    "    spe.hyperglycemia.nocovar$P[i]       <- spe_i$p.value                                     # P value from Wilcoxon rank-sum test\n",
    "}\n",
    "\n",
    "spe.hyperglycemia.nocovar$Difference <- spe.hyperglycemia.nocovar$Mean_1 - spe.hyperglycemia.nocovar$Mean_0 # Get differences between group 0 and 1\n",
    "spe.hyperglycemia.nocovar$FDR        <- p.adjust(spe.hyperglycemia.nocovar$P, method = \"fdr\")               # Get adjusted P value\n",
    "spe.hyperglycemia.nocovar            <- dplyr::arrange(spe.hyperglycemia.nocovar, P)                        # Sort result table by raw P value\n",
    "\n",
    "if(!dir.exists(\"associationNoCovar\")){dir.create(\"associationNoCovar\")}\n",
    "write.table(spe.hyperglycemia.nocovar, \"associationNoCovar/hyperglycemia.differential.species.wilcoxon.tsv\", sep = \"\\t\", row.names = F, col.names = T, quote = F)\n",
    "\n",
    "# Association between species abundance and continuous variables\n",
    "spe.glucose.nocovar <- as.data.frame(matrix(NA, ncol = 3, nrow = ncol(mbio.clr.filtered)))\n",
    "colnames(spe.glucose.nocovar)<-c(\"Species\", \"R\", \"P\")\n",
    "\n",
    "for (i in 1:ncol(mbio.clr.filtered)) {\n",
    "    spe_i <- cor.test(mbio.clr.filtered[ ,i], phen$Glucose, method = \"spearman\") # Spearman correlation analysis\n",
    "    \n",
    "    spe.glucose.nocovar$Species[i] <- colnames(mbio.clr.filtered)[i]             # Species name\n",
    "    spe.glucose.nocovar$R[i]       <- spe_i$estimate                             # R value (correlation strength) from Spearman correlation\n",
    "    spe.glucose.nocovar$P[i]       <- spe_i$p.value                              # P value from Spearman correlation\n",
    "}\n",
    "\n",
    "spe.glucose.nocovar$FDR <- p.adjust(spe.glucose.nocovar$P, method = \"fdr\") # Get adjusted P value\n",
    "spe.glucose.nocovar     <- dplyr::arrange(spe.glucose.nocovar, P)          # Sort result table by raw P value\n",
    "\n",
    "if(!dir.exists(\"associationNoCovar\")){dir.create(\"associationNoCovar\")}\n",
    "write.table(spe.glucose.nocovar, \"associationNoCovar/Glucose.species.spearman.tsv\", sep = \"\\t\", row.names = F, col.names = T, quote = F)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cb2d5968-0d3c-4645-b7c1-d9f0c428c752",
   "metadata": {},
   "source": [
    "### 5.2 Correlation between species abundance and phenotypes (with covariates)\n",
    "\n",
    "We use **linear regression** (when y is a continuous variable) or **logistic regression** (when y is a binary variable 0/1) to conduct association analysis, and add potential confounding factors in the models as covariates to remove the confounding effect. The common confounding factor in microbiome study is sequencing depth; In human microbiome study, we also need to consider age and gender as potential confounding factors."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "57f6dcbb-45b1-4aab-b087-6b02f2d280d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Association between species abundance and binary variables\n",
    "spe.hyperglycemia <- as.data.frame(matrix(NA, ncol = 3, nrow = ncol(mbio.clr.filtered)))\n",
    "colnames(spe.hyperglycemia)<-c(\"Species\", \"Beta\", \"P\")\n",
    "\n",
    "for (i in 1:ncol(mbio.clr.filtered)) {\n",
    "    spe_i <- glm(as.factor(Hyperglycemia) ~ Age+Gender+CleanReadCount+mbio.clr.filtered[,i], data = phen, family = \"binomial\") # logistic regresion analysis\n",
    "    spe_i_summ <- summary(spe_i)                                                                                # Summary the result to get P value\n",
    "    \n",
    "    spe.hyperglycemia$Species[i] <- colnames(mbio.clr.filtered)[i]  # Species name\n",
    "    spe.hyperglycemia$Beta[i]    <- spe_i_summ$coefficients[5,1]    # Beta coefficient from logistic regression\n",
    "    spe.hyperglycemia$P[i]       <- spe_i_summ$coefficients[5,4]    # P value from logistic regression\n",
    "}\n",
    "\n",
    "spe.hyperglycemia$FDR <- p.adjust(spe.hyperglycemia$P, method = \"fdr\") # Get adjusted P value\n",
    "spe.hyperglycemia     <- dplyr::arrange(spe.hyperglycemia, P)          # Sort result table by raw P value\n",
    "\n",
    "if(!dir.exists(\"associationCovar\")){dir.create(\"associationCovar\")}\n",
    "write.table(spe.hyperglycemia, \"associationCovar/Hyperglycemia.species.logisticReg.tsv\", sep = \"\\t\", row.names = F, col.names = T, quote = F)\n",
    "\n",
    "\n",
    "# Association between species abundance and continuous variables\n",
    "spe.glucose <- as.data.frame(matrix(NA, ncol = 3, nrow = ncol(mbio.clr.filtered)))\n",
    "colnames(spe.glucose)<-c(\"Species\", \"Beta\", \"P\")\n",
    "\n",
    "for (i in 1:ncol(mbio.clr.filtered)) {\n",
    "    spe_i <- lm(Glucose ~ Age+Gender+CleanReadCount+mbio.clr.filtered[,i], data = phen) # linear regresion analysis\n",
    "    spe_i_summ <- summary(spe_i)                                         # Summary the result to get P value\n",
    "    \n",
    "    spe.glucose$Species[i] <- colnames(mbio.clr.filtered)[i] # Species name\n",
    "    spe.glucose$Beta[i]    <- spe_i_summ$coefficients[5,1]   # Beta coefficient from linear regression\n",
    "    spe.glucose$P[i]       <- spe_i_summ$coefficients[5,4]   # P value from linear regression\n",
    "}\n",
    "\n",
    "spe.glucose$FDR <- p.adjust(spe.glucose$P, method = \"fdr\") # Get adjusted P value\n",
    "spe.glucose     <- dplyr::arrange(spe.glucose, P)          # Sort result table by raw P value\n",
    "\n",
    "if(!dir.exists(\"associationCovar\")){dir.create(\"associationCovar\")}\n",
    "write.table(spe.glucose, \"associationCovar/glucose.species.linearReg.tsv\", sep = \"\\t\", row.names = F, col.names = T, quote = F)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f7587090-dad1-42af-a23d-a7ce6d8f70b1",
   "metadata": {},
   "source": [
    "## 6 Mediation analysis\n",
    "\n",
    "**Mediation analysis** is a statistic technique which can help us **infer the causal relationship or regulatory relationship between multiple factors**. For instance, in the previous result, we can see that the bacterial species *Eubacterium eligens* showed the most significant and negative association with glucose, we would like to know if the *Eubacterium eligen* can be used as a target to decrease glucose using lifestyle intervention (e.g. eat more fruit), then we can test the regulation relationship between fruit intake frequency, *Eubacterium eligen* and glucose."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a9c575f1-3c2f-4b26-b576-7ee868761640",
   "metadata": {},
   "outputs": [],
   "source": [
    "Microbe<-mbio.clr.filtered[,grep(\"s__Eubacterium_eligens\", colnames(mbio.clr.filtered))]\n",
    "mediation.input<-data.frame(phen, Microbe)\n",
    "\n",
    "# Check independent variable (FruitIntakeFrequency) to mediator (Microbe)\n",
    "fit.mediator <- lm(Microbe~Age+Gender+CleanReadCount+FruitIntakeFrequency, data = mediation.input)\n",
    "summary(fit.mediator)\n",
    "# Check mediator (Microbe) to outcome (Glucose)\n",
    "fit.dv <- lm(Glucose~Age+Gender+CleanReadCount+FruitIntakeFrequency+Microbe, data = mediation.input)\n",
    "summary(fit.dv)\n",
    "\n",
    "# Mediation analysis\n",
    "mediation.res <- mediate(fit.mediator, fit.dv, treat='FruitIntakeFrequency', mediator='Microbe', boot=T)\n",
    "summary(mediation.res)a\n",
    "\n",
    "# ACME (average causal mediation effect) represents the indirect effect of independent factor (fruit intake frequency) on dependent factor or response factor (glucose) that goes through the mediator (Eubacterium eligen), this is the key parameter we are going to report in our study.\n",
    "# ADE (average direct effect) represents the direct effect of independent factor on dependent factor\n",
    "# Total Effect means the sum of direct and indirect effect\n",
    "# Prop. Mediated describes the proportion of indirect effect in total effect, this is also a parameter we are interested in."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.2.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
