{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b179745a-fc41-4149-a471-08efa6ee0dc7",
   "metadata": {},
   "source": [
    "# BioSB Computational Metagenomics Course"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d1e54d30-eca9-44c2-9fee-6e1f074b0dc4",
   "metadata": {},
   "source": [
    "### Practicals for metabolic modeling at the community level\n",
    "\n",
    "For the following practicals we will use python to explore, simulate, and play with metabolic models in a community context.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "595c66e7-e7c9-4fc2-8674-99b630d2c681",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "[cobrapy](https://cobrapy.readthedocs.io/en/latest/index.html#) is one of the best ways to gain \"programatic\" access to genome-scale metabolic models (GSMMs). \n",
    "\n",
    "Before going into advanced applications, we give a very brief introduction to some of its commom operations.\n",
    "\n",
    "Metabolic reconstructions are commonly stored as .xml files that follow the standard from the Systems Biology Markup Language ([sbml](https://en.wikipedia.org/wiki/SBML#:~:text=The%20Systems%20Biology%20Markup%20Language,community%20of%20users%20and%20developers.)). The models that we are going to use in the tutorial are available in the folder 'sbmlModels'.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "13def21f-dcfc-482c-8617-6372a4dc04c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "import cobra"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d31309a7-056e-4ae6-8113-49ee4edf1557",
   "metadata": {},
   "source": [
    "We demonstrate some of the basic functionalities of cobrapy. After this introduction, you should be able to:\n",
    " \n",
    "1) Load a model object; \n",
    "    &nbsp;\n",
    "2) Manipulate metabolites, reactions, and genes objects;\n",
    "    &nbsp;\n",
    "3) Solve the model with linear programing, using flux balance analysis (FBA);\n",
    "    &nbsp;\n",
    "4) Change the composition of the external environment.\n",
    "&nbsp;\n",
    "&nbsp;\n",
    "\n",
    "**Note**: in this tutorial we often refer to 'objects' such as 'model object', 'reaction object', etc. These are python classes, which contain a number of specific attributes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "af0ef693-da77-4ad1-85ac-cfa4e82dd0b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "#import cobra and set the solver to the free glpk\n",
    "import cobra\n",
    "cobra_config = cobra.Configuration()\n",
    "cobra_config.solver = 'glpk'\n",
    "\n",
    "\n",
    "#import other packages \n",
    "import numpy as np\n",
    "import os\n",
    "\n",
    "#folder with models\n",
    "modelFolder = '/data/precomputed/sbmlModels'\n",
    "\n",
    "#get a list with the models that are available in the folder\n",
    "print(os.listdir(modelFolder))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "328448ab-c47a-4213-b3f7-19c2165daa6a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#create a model object by loading an SBML file\n",
    "model = cobra.io.read_sbml_model(os.path.join(modelFolder,'acetogen_toy_model.xml' ))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "354a7cd7-245c-45fd-b971-8b0ddd59bfa3",
   "metadata": {},
   "source": [
    "#### Model\n",
    "\n",
    "The model object is a python class with all the relevant attributes to manipulate a GSMM.\n",
    "\n",
    "To see all the attributes...\n",
    "\n",
    "```python\n",
    "\n",
    "for i in dir(model):\n",
    "    if i[0]!='_':\n",
    "        print(i)\n",
    "```\n",
    "\n",
    "\n",
    "Beyond metabolites, reactions, and genes that we'll explain with more detail below, some of the relevant attributes to keep in mind are:\n",
    "\n",
    "- id \n",
    "contains the model's id\n",
    " `print(model.id)`\n",
    "\n",
    "- compartments\n",
    "contains the model's compartments and the 'key' by which they are referred to. For instance 'c' represents that cytoplasm and 'e' the external environment (or periplasm)\n",
    "`print(model.compartments)`\n",
    "\n",
    "\n",
    "Notice that model compartment dictionary is only intended for seeing the model compartments.\n",
    "\n",
    "`model.compartments['c'] = 'Cytoplasm'\n",
    "print(model.compartments)`\n",
    "\n",
    "To change add a name to the compartments, use:\n",
    "\n",
    "`model._compartments['c'] = 'Cytoplasm'\n",
    "print(model.compartments)`\n",
    "\n",
    "\n",
    "\n",
    "- medium\n",
    "contains the information about the concentration of metabolites that are available to the model from the external compartment (media metabolites). More information about this in the section about changing the environment.\n",
    "`print(model.medium)`\n",
    "\n",
    "\n",
    "- exchanges\n",
    "\n",
    "A list with reaction objects for the metabolites that could potentially be exchanged (in or out) by the model. \n",
    "`print(model.exchanges)`\n",
    "\n",
    "Notice that the list contains the actual reaction objects, not their ids.\n",
    "\n",
    "- sinks\n",
    "\n",
    "Similar to the exchanges, is a list with the sink reaction objects.\n",
    "`print(model.sinks)`"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b88d20b2-f866-4a59-b907-456ad553d41a",
   "metadata": {},
   "source": [
    "#### Metabolites\n",
    "A list that contains an object for each metabolite in the model.\n",
    "\n",
    "There are different functions to retrieve the desired metabolite object from the model, according to some query criteria:\n",
    "\n",
    "`model.metabolites.get_by_id` , `model.metabolites.get_by_any`, `model.metabolites.has_id`\n",
    "\n",
    "The most important attributes of a metabolite object are: \n",
    "\n",
    "- id - allows one to access the model object. \n",
    "```\n",
    "water = model.metabolites.cpd00001_c\n",
    "\n",
    "```\n",
    "\n",
    "if the database used contains ids that are not suitable as python variable names (such as the BIGG ids), the metabolite can be accessed with the `get_by_id` function of the `model.metabolites` object.\n",
    "\n",
    "```\n",
    "water = model.metabolites.get_by_id('cpd00001_c')\n",
    "```\n",
    "\n",
    "- name\n",
    "`print(water.name)`\n",
    "\n",
    "- formula\n",
    "`print(water.formula)`\n",
    "\n",
    "- elements\n",
    "`print(water.elements)`\n",
    "\n",
    "- charge\n",
    "`print(water.charge)`\n",
    "\n",
    "- compartment\n",
    "A metabolite object is specific for a single compartment. If the same metabolite occurs on different compartments, then new objects need to be made for each compartment that it occurs. \n",
    "\n",
    "The metabolite ids commonly have a '_c' or '_e' as a suffix, to indicate their compartments.\n",
    "\n",
    "`print(water.compartment)`\n",
    "\n",
    "- reactions\n",
    "list of reaction objects that the metabolite is a reactant or product.\n",
    "\n",
    "`print(water.reactions)`"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "706cc35e-f309-4fb6-8aea-065b94e64978",
   "metadata": {},
   "source": [
    "#### Reactions\n",
    "Also an attribute of the model object. Consists of a list containing a reaction object for each of the model's reactions.\n",
    "\n",
    "`print(len(model.reactions))`\n",
    "\n",
    "Similar to the metabolite objects, reaction objects may be accessed directly by their ids or with the special functions:\n",
    "\n",
    "`model.reactions.get_by_id` , `model.reactions.get_by_any`, `model.reactions.has_id`\n",
    "\n",
    "\n",
    "`atpase = model.reactions.rxn08173`\n",
    "\n",
    "or\n",
    "\n",
    "`atpase = model.reactions.get_by_id('rxn08173')`\n",
    "\n",
    "Some of the most important attributes of the reaction object are:\n",
    "\n",
    "- name\n",
    "`print(atpase.name)`\n",
    "\n",
    "- compartments\n",
    "while metabolites are present on a single compartments, reactions may occur across compartments.\n",
    "`print(atpase.compartments)`\n",
    "\n",
    "- metabolites\n",
    "A dictionary where the keys are the metabolite objects that participate in the reactions and the values are their stoichiometries. Reactants and products have, respectively, negative and positive values\n",
    "\n",
    "`print(atpase.metabolites)`\n",
    "\n",
    "- reactants\n",
    "A list with the metabolite objects of the reactants.\n",
    "`print(atpase.reactants)`\n",
    "\n",
    "- products\n",
    "A list with the metabolite objects of the products.\n",
    "`print(atpase.products)`\n",
    "\n",
    "- build_reaction_string\n",
    "`print(atpase.build_reaction_string())`\n",
    "\n",
    "To see the metabolite names:\n",
    "`print(atpase.build_reaction_string(use_metabolite_names=1))`\n",
    "\n",
    "\n",
    "\n",
    "- lower_bound\n",
    "The lowest flux value that is allowed to go through the reaction when solving the model.\n",
    "`print(atpase.lower_bound)`\n",
    "\n",
    "- upper_bound\n",
    "The highest flux value that is allowed to go through the reaction when solving the model.\n",
    "`print(atpase.upper_bound)`\n",
    "\n",
    "**note:** the upper and lower bounds control the reaction directionality. \n",
    "\n",
    "```python\n",
    "\n",
    "print(atpase.build_reaction_string(use_metabolite_names=1))\n",
    "atpase.lower_bound = 0\n",
    "print(atpase.build_reaction_string(use_metabolite_names=1))\n",
    "\n",
    "```\n",
    "\n",
    "They are also the easiest way to 'knockout' a reaction.\n",
    "\n",
    "```python\n",
    "\n",
    "atpase.lower_bound = 0\n",
    "atpase.upper_bound = 0\n",
    "\n",
    "```\n",
    "\n",
    "Or to 'force' flux through a reaction (but in many cases would make the model unfeasible)\n",
    "\n",
    "```python\n",
    "\n",
    "atpase.lower_bound = 10\n",
    "atpase.upper_bound = 1000\n",
    "\n",
    "```\n",
    "\n",
    "- flux\n",
    "The flux that goes through a reaction is only accessible once the model is solved. \n",
    "\n",
    "\n",
    "`print(atpase.flux)`\n",
    "\n",
    "##### Types of reactions\n",
    "\n",
    "We can distinguish reactions based on the compartments of the metabolites\n",
    "\n",
    "\n",
    "- Internal\n",
    "\n",
    "Reactions that occur only in the cytosol. E.g.:\n",
    "\n",
    "```python\n",
    "\n",
    "fdh = model.reactions.FDH\n",
    "\n",
    "print(fdh.compartments)\n",
    "\n",
    "```\n",
    "\n",
    "\n",
    "- Objective\n",
    "The objective reaction that is 'maximized' (or minimized) by the model.\n",
    "`print(model.objective_direction)`\n",
    "\n",
    "```python\n",
    "\n",
    "for i in model.reactions:\n",
    "    if i.objective_coefficient==1:\n",
    "        print(i.id)`\n",
    "\n",
    "\n",
    "```\n",
    "\n",
    "- Transport\n",
    "\n",
    "Reactions that take extracellular metabolites and transport them into the cell (i.e. convert them into cytosol metabolites)\n",
    "\n",
    "```python\n",
    "\n",
    "gluT = model.reactions.rxn05147\n",
    "\n",
    "print(gluT.build_reaction_string(use_metabolite_names=0))\n",
    "\n",
    "print(gluT.upper_bound)\n",
    "\n",
    "print(gluT.lower_bound)\n",
    "\n",
    "```\n",
    "\n",
    "\n",
    "Notice that 'cpd00027_e' is converted to 'cpd00027_c'.\n",
    "\n",
    "- Exchange\n",
    "\n",
    "Reactions that make a metabolite available in the extracellular space. From the extracellular space it could be taken up by a transporter.\n",
    "\n",
    "```python\n",
    "\n",
    "print([i.id for i in model.exchanges])\n",
    "\n",
    "gluExch = model.reactions.EX_cpd00027_e\n",
    "\n",
    "print(gluExch.build_reaction_string(use_metabolite_names=1))\n",
    "\n",
    "print(gluExch.lower_bound)\n",
    "\n",
    "print(gluExch.upper_bound)\n",
    "\n",
    "```\n",
    "\n",
    "Notice that the inward flux is negative. If the flux is positive, it means the metabolite is getting secreted. \n",
    "\n",
    "\n",
    "- Sink \n",
    "\n",
    "Reactions that eliminate a dead-end metabolite (i.e. a metabolite that is not consumed by any further reaction and is not secreted)\n",
    "\n",
    "```python\n",
    "\n",
    "piSk = model.reactions.piSink\n",
    "\n",
    "piSk.build_reaction_string(use_metabolite_names=1)\n",
    "\n",
    "print(piSk.lower_bound)\n",
    "\n",
    "print(piSk.upper_bound)`\n",
    "\n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "26f60378-0b47-426b-a8ef-e2b84bc43702",
   "metadata": {},
   "source": [
    "## Task: Model manipulation 1 (5-10 min)\n",
    "\n",
    "Make a summary table containing all the reactions in the toy models with the following fields:\n",
    "\n",
    "- Id\n",
    "\n",
    "- Name\n",
    "- Reaction string with metabolite names\n",
    "- Lower bound\n",
    "- Upper bound\n",
    "- Directionality\n",
    "- Type (internal, objective, transport, sink, exchange)\n",
    "- SugarFermenter (1 if present in the sugar fermenter, 0 otherwise)\n",
    "- ButyrateProducer (1 if present in the butyrate producer, 0 otherwise)\n",
    "- Acetogen (1 if present in the acetogen, 0 otherwise)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e87a7e6d-fe78-4090-8187-e15f19e5bffe",
   "metadata": {},
   "source": [
    "#### genes\n",
    "\n",
    "The model object also contais genes.\n",
    "\n",
    "```python\n",
    "#change to the 'full' model\n",
    "\n",
    "model = cobra.io.read_sbml_model(os.path.join(modelFolder,'acetogen_full_model.xml' ))\n",
    "\n",
    "\n",
    "print(model.genes)\n",
    "\n",
    "gene1 = model.genes[0]\n",
    "\n",
    "print(gene1.name)\n",
    "\n",
    "print(gene1.id)\n",
    "\n",
    "print(gene1.reactions)\n",
    "\n",
    "```\n",
    "Note that a single gene can explain several reactions and a reaction may be explained by one or more genes. Boolean rules may be applied to explain if the reaction is an isozyme (OR) or part of a complex (AND).\n",
    "\n",
    "```python\n",
    "for i in gene1.reactions:\n",
    "    print(i.gene_reaction_rule)\n",
    "\n",
    "```\n",
    "\n",
    "Gapfilled reactions and some others (sink, exchange) do not have genes associated to them:\n",
    "\n",
    "```python\n",
    "geneless = [i.id for i in model.reactions if i.genes == frozenset()]\n",
    "\n",
    "```\n",
    "\n",
    "\n",
    "```python\n",
    "#change back to the 'toy' model\n",
    "\n",
    "model = cobra.io.read_sbml_model(os.path.join(modelFolder,'acetogen_toy_model.xml' ))\n",
    "\n",
    "```\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "**note** the toy models do not contain any genes associated to them.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "676590e9-84e5-44fd-a660-7c618569225f",
   "metadata": {},
   "source": [
    "#### solving the model\n",
    "\n",
    "We can solve the model using FBA. \n",
    "\n",
    "```python\n",
    "obj = model.optimize()\n",
    "\n",
    "print(obj.objective_value)\n",
    "\n",
    "print(obj.fluxes)\n",
    "\n",
    "```\n",
    "\n",
    "For algorithms that require the objective_value in a fast way, use:\n",
    "\n",
    "```python\n",
    "objCoef = model.slim_optimize()\n",
    "\n",
    "print(objCoef)\n",
    "\n",
    "```\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c1d7a8c9-9f17-4add-a074-fa1074b1cc8b",
   "metadata": {},
   "source": [
    "## Task: model manipulation 2 (10-15 min)\n",
    "\n",
    "- Find reactions that are essential for the model to produce biomass\n",
    "\n",
    "- Find reactions that are only essential in pairs\n",
    "\n",
    "- Repeat the above, but now find reaction and reaction pairs that reduce the modelâ€™s biomass by up to an arbitrary threshold (i.e. 10%)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "44ae5707-3e1a-4364-b6a7-aed6cf7aa43c",
   "metadata": {},
   "source": [
    "#### Changing the environment\n",
    "\n",
    "The metabolites are made available to the models through the exchange reactions (see above).\n",
    "\n",
    "To 'control' the composition, set the lower bound. But take into account that these are constrained-based models. To 'force' the models to consume a defined amount, set the lower and upper bound to this amount (this often leads to unfeasible models)\n",
    "\n",
    "```python\n",
    "#provide a rich media\n",
    "for i in model.exchanges:\n",
    "    i.lower_bound = -1000\n",
    "\n",
    "obj = model.optimize()\n",
    "\n",
    "print(obj.objective_value)\n",
    "\n",
    "\n",
    "#remove glucose \n",
    "model.reactions.EX_cpd00027_e.lower_bound=0\n",
    "\n",
    "obj = model.optimize()\n",
    "\n",
    "print(obj.objective_value)\n",
    "\n",
    "#force a glucose flux\n",
    "\n",
    "model.reactions.EX_cpd00027_e.lower_bound=-30\n",
    "model.reactions.EX_cpd00027_e.upper_bound=-30\n",
    "obj = model.optimize()\n",
    "\n",
    "print(obj.objective_value)\n",
    "\n",
    "\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d7585a36-731d-4576-83ba-c86fca98f344",
   "metadata": {},
   "source": [
    "## Task: model manipulation 3 (10-15 min)\n",
    "\n",
    "- Generate random environment compositions and associate them to the production of biomass by the three toy models\n",
    "\n",
    "- Compare their growth rate distributions across these random environments (making box plots should be enough)\n",
    "\n",
    "- Make the same comparison without glucose\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
